### PR TITLE
gsoc: explain: Add explain plugin

### PR DESCRIPTION
**Pull request by meetmangukiya**
Closes https://github.com/coala/corobo/issues/7

### CODE REVIEW

File: plugins/explain.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                  'discussion, please add information [at our GitLab wiki page]'
+                  '(https://github.com/coala/coala/wiki/GitLab).',
+        'google': 'Hey. This message was triggered because someone was too '
+                  "lazy to type this *again*. Don't take it personally. "
+                  'Please.\n\nWe all got to learn this: *use google*. Or '
+                  'duckduckgo. Anything. The search engine that earned your '
+                  'trust. You got a build error? Search for the first red '
+                  'thing and google it. You got an exception? *Read the '
+                  'message.* Search it. *Think.*\n\nKeep this in mind: *You*'
+                  'are sitting in front of the problem, not we. You will have '
```
Comment (Udayan12167) [inline plugins/explain.py]: `*not us`

File: plugins/explain.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                  "a much easier time solving it. That's why you should try "
```
Comment (Udayan12167) [inline plugins/explain.py]: Double quotes.

File: plugins/explain.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                  'help you understand the basics of how it works and what you'
+                  "should be doing.\n\nIf you're just looking to fix an issue "
+                  'very quickly and not interested in contributing to coala '
+                  'long term, we can fix it up for you - just tell us :).',
+        'cep': "At coala we're using [cEP's (coala Enhancement Proposals)]"
+               '(http://coala.io/cep) to define major design decisions - '
+               "they're a bit like PEP's but not quite as extensive and "
+               'obviously written with a lower case c.',
+        'gitlab': 'We are currently evaluating on if we want to use GitLab for'
+                  "code hosting. That's why some repositories are already on "
```
Comment (Udayan12167) [inline plugins/explain.py]: Double quotes.

File: plugins/explain.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    Explain various terms
+    """
+
+    MSGS = {
+        'review': 'After creating your `Pull Request`, it is under the review '
+                  'process. This can be deduced from the `process/pending '
+                  'review` label. Now you have to wait for the reviewers to '
+                  'review your PR. You should *not* ask for reviews on our '
+                  'Gitter channel - we review those PRs continuously.\n\n'
+                  "We're usually swamped with reviews, while you are waiting "
```
Comment (Udayan12167) [inline plugins/explain.py]: Double quotes

File: plugins/explain.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                  "**please review other people's PRs** at [coala.io/review]"
```
Comment (Udayan12167) [inline plugins/explain.py]: Double quotes

File: plugins/explain.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                  '(https://coala.io/review): that helps you and will make '
+                  'your review happen faster as well. As a rule of thumb, '
+                  '*for every review you receive, give at least one review '
+                  'to someone else!*\n\nFor a good review, look at every '
+                  'commit on its own and place `<sha> is ready` or `<sha> '
+                  'needs work` comments on the pull request, be sure to '
+                  "remove other spacing like tabs. If you're done with a "
```
Comment (Udayan12167) [inline plugins/explain.py]: Dq

File: plugins/explain.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        'fixes': 'We use bug prediction in coala which relies on the `Fixes` '
+                 'keyword in commit messages. To get good results from that '
+                 'we need to use `Fixes` for bugfix issues instead of '
+                 '`Closes`. (See also [the commit message docs]'
+                 '(https://coala.io/commit).) To change your message you '
+                 'just use `git commit --amend` and then `git push --force` '
+                 'the new commit to replace the old one.',
+        'commit message': 'To change your message you just use `git commit '
+                          '--amend` and then `git push --force` the new '
+                          "commit to replace the old one.\n\nIf you're just "
```
Comment (Udayan12167) [inline plugins/explain.py]: Dq

File: plugins/explain.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                          'looking to fix an issue very quickly and not '
+                          'interested in contributing to coala long term, we '
+                          'can fix up the message for you - just tell us :).',
+        'rebase': 'It looks like your PR is out of date and needs a rebase.'
+                  '\n\n[This page](https://coala.io/git) may help you to get '
+                  'started on this. We also have [a quick video tutorial on '
+                  'how to rebase](https://asciinema.org/a/78683). That should '
+                  'help you understand the basics of how it works and what you'
+                  "should be doing.\n\nIf you're just looking to fix an issue "
```
Comment (Udayan12167) [inline plugins/explain.py]: Dq

File: plugins/explain.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                  'very quickly and not interested in contributing to coala '
+                  'long term, we can fix it up for you - just tell us :).',
+        'cep': "At coala we're using [cEP's (coala Enhancement Proposals)]"
```
Comment (Udayan12167) [inline plugins/explain.py]: Dq

File: plugins/explain.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+               '(http://coala.io/cep) to define major design decisions - '
+               "they're a bit like PEP's but not quite as extensive and "
```
Comment (Udayan12167) [inline plugins/explain.py]: Dq

File: plugins/explain.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+               'obviously written with a lower case c.',
+        'gitlab': 'We are currently evaluating on if we want to use GitLab for'
+                  "code hosting. That's why some repositories are already on "
+                  'GitLab, if you want to participate in the migration '
+                  'discussion, please add information [at our GitLab wiki page]'
+                  '(https://github.com/coala/coala/wiki/GitLab).',
+        'google': 'Hey. This message was triggered because someone was too '
+                  "lazy to type this *again*. Don't take it personally. "
```
Comment (Udayan12167) [inline plugins/explain.py]: Dq

File: plugins/explain.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+                  'review` label. Now you have to wait for the reviewers to '
+                  'review your PR. You should *not* ask for reviews on our '
+                  'Gitter channel - we review those PRs continuously.\n\n'
+                  "We're usually swamped with reviews, while you are waiting "
+                  "**please review other people's PRs** at [coala.io/review]"
+                  '(https://coala.io/review): that helps you and will make '
+                  'your review happen faster as well. As a rule of thumb, '
+                  '*for every review you receive, give at least one review '
+                  'to someone else!*\n\nFor a good review, look at every '
+                  'commit on its own and place `<sha> is ready` or `<sha> '
```
Comment (Udayan12167) [inline plugins/explain.py]: `ack <sha>`?

File: plugins/explain.py (hunk 2 above)
Comment (meetmangukiya) [inline plugins/explain.py]: Single quote in the quote, hence the double quote, I'd rather not have an escaped single quote

Comment (meetmangukiya) [inline plugins/explain.py]: Also, single quotes is enforced by `QuotesBear` and you wouldn't have to manually check the quotes now :wink:

File: plugins/explain.py (hunk 3 above)
Comment (meetmangukiya) [inline plugins/explain.py]: same

File: plugins/explain.py (hunk 4 above)
Comment (meetmangukiya) [inline plugins/explain.py]: same

File: plugins/explain.py (hunk 5 above)
Comment (meetmangukiya) [inline plugins/explain.py]: same

File: plugins/explain.py (hunk 6 above)
Comment (meetmangukiya) [inline plugins/explain.py]: same

File: plugins/explain.py (hunk 7 above)
Comment (meetmangukiya) [inline plugins/explain.py]: same

File: plugins/explain.py (hunk 8 above)
Comment (meetmangukiya) [inline plugins/explain.py]: same

File: plugins/explain.py (hunk 9 above)
Comment (meetmangukiya) [inline plugins/explain.py]: same

File: plugins/explain.py (hunk 10 above)
Comment (meetmangukiya) [inline plugins/explain.py]: same

File: plugins/explain.py (hunk 11 above)
Comment (meetmangukiya) [inline plugins/explain.py]: same

File: plugins/explain.py (hunk 12 above)
Comment (meetmangukiya) [inline plugins/explain.py]: hmm.. no idea why this, I copied all of the contents/messages from old cobot explain command

Comment (meetmangukiya) [inline plugins/explain.py]: @sils we are changing `sha is ready`, to `ack sha` and `needs work` to `unack sha`, no idea why is it written this way in old cobot.

**Comment by Udayan12167** [comment]
ack https://github.com/coala/corobo/pull/59/commits/ce57de8bb5d5634b996455655289c805afc7616e

**Comment by meetmangukiya** [comment]
@rultor merge

**Pull request merged**